File: broker.js
#!/usr/bin/env nodejs

// MQTT to Telegram Message Broker
// Copyright (C) 2016  Lucas Teske

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

const { QLog, undefinedOrNull } = require('quanto-commons');

const mqtt = require('mqtt')
const TelegramBot = require('node-telegram-bot-api');

const telegramBotToken = process.env['telegram_bot_token'];
const defaultGroupId = !undefinedOrNull(process.env['telegram_group_id']) ? parseInt(process.env['telegram_group_id'], 10) : null;
const mqttHost = process.env['mqtt_server'];
const telegramAdminId = process.env['telegram_admin'];
const msgToName = process.env['mqtt_message_to'];

const groupToTopic = process.env['group_to_topic'];

const GlobalLog = QLog.scope('Global');
const TelLog = QLog.scope('Telegram');
const MQTTLog = QLog.scope('MQTT');

MQTTLog.enableLogs(['debug', 'warn']);
GlobalLog.enableLogs(['warn']);
TelLog.enableLogs(['warn']);

GlobalLog.headPadding = 30;
TelLog.headPadding = 30;
MQTTLog.headPadding = 30;

// region Variable Check
if (undefinedOrNull(telegramBotToken)) {
  GlobalLog.error(`Telegram Bot Token was not defined! Please define at environment variable ${'telegram_bot_token'.warn.bold}`);
}
if (undefinedOrNull(defaultGroupId)) {
  GlobalLog.error(`Telegram Group ID was not defined! Please define at environment variable ${'telegram_group_id'.warn.bold}`);
}
if (undefinedOrNull(mqttHost)) {
  GlobalLog.error(`MQTT Server was not defined! Please define at environment variable ${'mqtt_server'.warn.bold}`);
}
if (undefinedOrNull(telegramAdminId)) {
  GlobalLog.warn(`Telegram Administrator ID not defined. Administrator will be disabled. Define at environment variable ${'telegram_admin'.warn.bold}`);
}
if (undefinedOrNull(groupToTopic)) {
  GlobalLog.error(`Group to Topic was not defined! Please define at environment variable ${'group_to_topic'.warn.bold}`);
  GlobalLog.warn(`Format: groupId:mqttTopic:messageTo;groupId2:mqttTopic2:messageTo2`);
}

if (undefinedOrNull(telegramBotToken) || undefinedOrNull(defaultGroupId) || undefinedOrNull(groupToTopic) || undefinedOrNull(mqttHost)) {
  GlobalLog.fatal('One or more environment variables not defined. Aborting...');
  process.exit(1);
}

// endregion
const groupMaps = {};
const topicMaps = {};
const topicToMap = {};

groupToTopic.split(';').forEach(m => {
  const z = m.split(':');
  const group = parseInt(z[0], 10);
  const topic = z[1];
  MQTTLog.info(`Mapping Telegram Group ${group.toString().white.bold} to MQTT Topic ${topic.white.bold}`);
  groupMaps[group] = topic;
  topicMaps[topic] = group;

  if (z.length > 2) {
    topicToMap[topic] = z[2];
  } else {
    MQTTLog.warn(`Topic ${topic} does not have the third argument which represents the message to.`);
  }
});

const bot = new TelegramBot(telegramBotToken, { polling: true });
const client  = mqtt.connect(`mqtt://${mqttHost}`)

client.on('connect', function () {
  MQTTLog.success('Connected');
  client.subscribe('presence');
  Object.keys(topicMaps).forEach(topic => {
    MQTTLog.info(`Subscribing topic ${topic}`);
    client.subscribe(topic);
  });
})

const opts = { parse_mode: 'Markdown' };

const doMessage = (topic, jsonData) => {
  try {
    const data = JSON.parse(jsonData);
    if (data.type === 'message') {
      const group = topicMaps[topic];
      if (undefinedOrNull(group)) {
        MQTTLog.warn(`Received message on ${topic} but no telegram channel associated.`);
        return;
      }

      if (!undefinedOrNull(data.message)) {
        MQTTLog.info(`[${group.toString().white}] ${(data.from||'Unknown').warn.bold}: ${data.message}`);
        bot.sendMessage(group, `*${data.from}*: ${data.message}`, opts);
      } else {
        MQTTLog.error(`Received Data without message: ${jsonData}`);
        client.publish(`${topic}_error`, 'Received a payload without message.');
      }
    } else {
      MQTTLog.info(`Receive ${data.type} message. ${jsonData}`);
    }
  } catch(e) {
    client.publish(`${topic}_error`, `There was an error parsing message from ${topic}: ${e}`);
  }
}

client.on('message', function (topic, message) {
  MQTTLog.debug(`Received Message on Topic ${topic}: ${message}`);
  doMessage(topic, message);
});

bot.on('message', function (msg) {
  // TelLog.debug(JSON.stringify(msg, null, 2));
  if (msg.chat.id !== msg.from.id) {
    TelLog.note(`[${msg.chat.title.white}(${msg.chat.id.toString().gray})] ${(msg.from.username || 'Unknown').warn.bold}: ${msg.text}`);
  } else {
    TelLog.note(`${(msg.from.username || 'Unknown').warn.bold}: ${msg.text}`);
  }

  const topic = groupMaps[msg.chat.id];

  if (!undefinedOrNull(topic)) {
    const topicTo = topicToMap[topic];
    TelLog.debug(`Redirecting message from Channel: ${msg.chat.title.white.bold}`);
    if (!undefinedOrNull(topicTo)) {
      client.publish(`${topic}_msg`, JSON.stringify({
        "sendmsg": true,
        "to": topicTo,
        "message": `${msg.from.first_name} ${msg.from.last_name}: ${msg.text}`,
      }));
    } else {
      TelLog.error('Received message but can\'t send because no msgToName defined!');
    }
  }
});

bot.on('channel_post', function (msg) {
  const topic = groupMaps[msg.chat.id];
  if (!undefinedOrNull(topic)) {
    TelLog.info(`Received From Channel: ${msg.text}`);
    const topicTo = topicToMap[topic];
    if (!undefinedOrNull(topicTo)) {
      client.publish(`${topic}_msg`, JSON.stringify({
        "sendmsg": true,
        "to": topicTo,
        "message": msg
      }));
    } else {
      TelLog.error('Received message but can\'t send because no msgToName defined!');
    }
  }
});

File: broker.go
package main

import (
	"encoding/json"
	"fmt"
	"github.com/eclipse/paho.mqtt.golang"
	"github.com/go-telegram-bot-api/telegram-bot-api"
	"github.com/quan-to/slog"
	"os"
	"os/signal"
	"strconv"
	"strings"
	"syscall"
	"time"
)

var (
	telegramBotToken = os.Getenv("telegram_bot_token")
	telegramAdminId  = os.Getenv("telegram_admin")
	groupToTopic     = os.Getenv("group_to_topic")
	mqttHost         = os.Getenv("mqtt_server")
)

var telLog = slog.Scope("Telegram")
var mqttLog = slog.Scope("MQTT")

var groupMaps = map[int64]string{}
var topicMaps = map[string]int64{}
var topicToMap = map[string]string{}

var telegramBot *tgbotapi.BotAPI
var mqttClient mqtt.Client

func doMessage(topic string, jsonData []byte) {
	defer func() {
		if r := recover(); r != nil {
			mqttLog.Error("Recovered from panic on doMessage.")
			mqttClient.Publish(fmt.Sprintf("%s_error", topic), 0, false, fmt.Sprintf("There was an error processing the message: recovered from panic"))
		}
	}()

	var data map[string]interface{}
	err := json.Unmarshal(jsonData, &data)
	if err != nil {
		mqttLog.Error("Received invalid JSON: %s", err)
		mqttClient.Publish(fmt.Sprintf("%s_error", topic), 0, false, fmt.Sprintf("There was an error processing the message: %s", err))
		return
	}

	t := data["type"].(string)

	if t == "message" {
		group, ok := topicMaps[topic]
		if !ok {
			mqttLog.Warn("Received message on topic %s but no telegram channel associated.", topic)
			return
		}

		if data["message"] != nil {
			from := "Unknown"
			if data["from"] != nil {
				from = data["from"].(string)
			}
			message := data["message"].(string)
			mqttLog.Info("[%d] %s: %s", group, from, message)

			msg := tgbotapi.NewMessage(group, fmt.Sprintf("*%s*: %s", from, message))
			msg.ParseMode = tgbotapi.ModeMarkdown

			_, err := telegramBot.Send(msg)
			if err != nil {
				telLog.Error("Error sending message to group %d: %s", group, err)
			}
		} else {
			mqttLog.Error("Received data without message: %s", string(jsonData))
			mqttClient.Publish(fmt.Sprintf("%s_error", topic), 0, false, fmt.Sprintf("Received data without message: %s", string(jsonData)))
		}
	} else {
		mqttLog.Info("Received message (%s): %s", t, string(jsonData))
	}
}

func CheckTelegramUpdates() {
	u := tgbotapi.NewUpdate(0)
	u.Timeout = 60

	updates, err := telegramBot.GetUpdatesChan(u)

	if err != nil {
		telLog.Error("Error fetching updates: %s", err)
		return
	}

	for update := range updates {
		if update.ChannelPost != nil {
			msg := update.ChannelPost

			from := msg.Chat.Title
			telLog.Info("%s: %s", from, msg.Text)

			topic, ok := groupMaps[msg.Chat.ID]

			if ok {
				topicTo, ok := topicToMap[topic]
				telLog.Debug("Redirecting message from Channel: %s", msg.Chat.Title)
				if ok {

					data := map[string]interface{}{
						"sendmsg": true,
						"to":      topicTo,
						"message": msg.Text,
					}

					jsonData, _ := json.Marshal(data)
					mqttLog.Debug("Publishing to %s_msg: %s", topic, string(jsonData))
					mqttClient.Publish(fmt.Sprintf("%s_msg", topic), 0, false, jsonData)
				} else {
					telLog.Error("Received message but can't send because no msgToName defined!")
				}
			}
		}

		if update.Message != nil { // ignore any non-Message Updates
			msg := update.Message

			from := msg.From.UserName
			if from == "" {
				from = "Unknown"
			}

			if msg.Chat.ID != int64(msg.From.ID) {
				telLog.Info("[%s(%d)] %s: %s", msg.Chat.Title, msg.Chat.ID, from, msg.Text)
			} else {
				telLog.Info("%s: %s", from, msg.Text)
			}

			topic, ok := groupMaps[msg.Chat.ID]

			if ok {
				topicTo, ok := topicToMap[topic]
				telLog.Debug("Redirecting message from User: %s", msg.Chat.Title)
				if ok {

					data := map[string]interface{}{
						"sendmsg": true,
						"to":      topicTo,
						"message": fmt.Sprintf("%s %s: %s", msg.From.FirstName, msg.From.LastName, msg.Text),
					}

					jsonData, _ := json.Marshal(data)
					mqttLog.Debug("Publishing to %s_msg: %s", topic, string(jsonData))
					mqttClient.Publish(fmt.Sprintf("%s_msg", topic), 0, false, jsonData)
				} else {
					telLog.Error("Received message but can't send because no msgToName defined!")
				}
			}
		}
	}
}

func main() {
	var err error

	if telegramBotToken == "" {
		slog.Error("Telegram Bot Token was not defined! Please define at environment variable \"telegram_bot_token\"")
	}

	if mqttHost == "" {
		slog.Error(`MQTT Server was not defined! Please define at environment variable 'mqtt_server'`)
	}

	if telegramAdminId == "" {
		slog.Warn(`Telegram Administrator ID not defined. Administrator will be disabled. Define at environment variable 'telegram_admin'`)
	}

	if groupToTopic == "" {
		slog.Error(`Group to Topic was not defined! Please define at environment variable 'group_to_topic'`)
		slog.Warn(`Format: groupId:mqttTopic:messageTo;groupId2:mqttTopic2:messageTo2`)
	}

	if telegramBotToken == "" || groupToTopic == "" || mqttHost == "" {
		slog.Fatal("One or more environment variables not defined. Aborting...")
	}

	groups := strings.Split(groupToTopic, ";")

	for _, m := range groups {
		z := strings.Split(m, ":")
		group, _ := strconv.ParseInt(z[0], 10, 64)
		topic := z[1]

		mqttLog.Info("Mapping Telegram Group %d to MQTT Topic %s", group, topic)

		groupMaps[group] = topic
		topicMaps[topic] = group

		if len(z) > 2 {
			topicToMap[topic] = z[2]
		} else {
			mqttLog.Warn("Topic %s does not have a third argument which represents the message to.")
		}
	}

	slog.Info("Starting")
	// region Telegram Bot Connect
	telegramBot, err = tgbotapi.NewBotAPI(telegramBotToken)
	if err != nil {
		telLog.Fatal(err)
	}

	telegramBot.Debug = true

	telLog.Info("Authorized on account %s", telegramBot.Self.UserName)
	// endregion
	// region MQTT
	opts := mqtt.NewClientOptions()
	opts.AddBroker(fmt.Sprintf("tcp://%s:1883", mqttHost))
	opts.SetDefaultPublishHandler(func(client mqtt.Client, message mqtt.Message) {
		mqttLog.Debug(`Received Message on Topic %s: %s`, message.Topic(), string(message.Payload()))
		doMessage(message.Topic(), message.Payload())
	})
	opts.SetPingTimeout(1 * time.Second)
	opts.SetKeepAlive(2 * time.Second)

	mqttClient = mqtt.NewClient(opts)
	if token := mqttClient.Connect(); token.Wait() && token.Error() != nil {
		mqttLog.Fatal(token.Error())
	}

	mqttLog.Info("Connected")

	token := mqttClient.Subscribe("presence", 0, nil)
	token.Wait()
	err = token.Error()
	if err != nil {
		mqttLog.Fatal("Error subscribing to %s: %s", "presence", err)
	}

	for k := range topicMaps {
		token := mqttClient.Subscribe(k, 0, nil)
		token.Wait()
		err = token.Error()
		if err != nil {
			mqttLog.Fatal("Error subscribing to %s: %s", k, err)
		}
	}
	// endregion

	c := make(chan os.Signal, 1)
	done := make(chan bool, 1)
	signal.Notify(c, os.Interrupt, syscall.SIGINT, syscall.SIGTERM, syscall.SIGABRT)

	go func() {
		sig := <-c
		slog.Warn("Received Signal %d", sig)
		done <- true
	}()

	tick := time.NewTicker(time.Second * 1)
	running := true

	slog.Info("Starting global loop")

	for running {
		select {
		case <-tick.C:
			CheckTelegramUpdates()
		case <-done:
			running = false
		}
	}
	slog.Info("MQTT Telegram Stopped")
}
